File: Heap_Sort/Ruby/Sweet-kid/heap_sort.rb
class HeapSort

  attr_reader :array
  def initialize(array)
    @array = array
    heap_sort()
  end

  def heap_sort
    build_max_heap()
    loop_over = (1..(@array.length - 1)).to_a.reverse
    heap_size = @array.length
    for i in loop_over
      temp = @array[ 0 ]
      @array[ 0 ] = @array[ i ]
      @array[ i ] = temp
      heap_size -= 1
      max_heapify( 0, heap_size )
    end
  end

  def build_max_heap()
    upper = @array.length / 2 - 1
    loop_over = ( 0..upper).to_a.reverse
    for i in loop_over
      max_heapify( i, @array.length )
    end
  end

  def max_heapify( i, size )
    l = left( i )
    r = right( i )

    largest = i
    if( l < size and @array[ l ] > @array[ i ] )
      largest = l
    end

    if( r < size and @array[ r ] > @array[ largest ] )
      largest = r
    end

    if( largest != i )
      temp = @array[ i ]
      @array[ i ] = @array[ largest ]
      @array[ largest ] = temp
      max_heapify( largest, size )
    end
  end

  def parent( i )
    return ( i + 1 ) / 2 - 1
  end

  def left( i )
    return 2 * i + 1
  end

  def right( i )
    return 2 * i + 2
  end

  def print_array
    for i in @array
      print i, " "
    end
    puts "\n"
  end

end

File: Heap_Sort/Python/Sweet-kid/heap_sort.py
#!/usr/bin/env python

def heap_sort( array ) :
    heap_size = len( array )
    build_max_heap( array )
    for i in range( len( array ) - 1, 0, -1 ):
        temp = array[ 0 ]
        array[ 0 ] = array[ i ]
        array[ i ] = temp
        heap_size -= 1
        max_heapify( array, 0, heap_size )

def build_max_heap( array ):
    for i in range( int( len(array) / 2 ) - 1, -1, -1 ):
        max_heapify( array, i, len( array ) )

def max_heapify( array, i, size ):
    l = left( i )
    r = right( i )
    largest = i;

    if( l < size and array[ l ] > array[ i ] ):
        largest = l

    if( r < size and array[ r ] > array[ largest ] ):
        largest = r

    if( largest != i ):
        temp = array[ i ]
        array[ i ] = array[ largest ]
        array[ largest ] = temp
        max_heapify( array, largest, size )

def parent( i ):
    return int( ( i + 1 ) / 2 ) - 1

def left( i ):
    return 2 * i + 1

def right( i ):
    return 2 * i + 2

def print_array( array ):
    for i in range( 0, len(array) ):
        print array[ i ],

    print "\n\n"
